### PR TITLE
feat: cancel ES request from metastore side if indexsearch get cancelled from the client

### PR DESCRIPTION
**Pull request by sumandas0**
> If frontend cancels Indexsearch we still process it in ES which blocks the future requests which we want to succeed, so we are now supporting with help with client, client will send us context id and sequence number for a search and will cancel the previous call if indexsearch cames with same sequence id and higher sequence number.

## Type of change
- [ ] Bug fix (fixes an issue)
- [x] New feature (adds functionality)

## Related issues

> Fix [#1]() 

### Development

- [x] Lint rules pass locally
- [x] Application changes have been tested thoroughly
- [ ] Automated tests covering modified code pass

### Security

- [ ] Security impact of change has been considered
- [ ] Code follows company security practices and guidelines

### Code review 

- [ ] Pull request has a descriptive title and context useful to a reviewer. Screenshots or screencasts are attached as necessary
- [ ] "Ready for review" label attached and reviewers assigned
- [ ] Changes have been reviewed by at least one other contributor
- [ ] Pull request linked to task tracker where applicable

### CODE REVIEW

File: graphdb/janus/src/main/java/org/apache/atlas/repository/graphdb/janus/SearchContextCache.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+    public static Integer getSequence(String key) {
+        return (Integer) searchContextSequenceCache.getIfPresent(key);
+    }
+
+    public static String get(String key){
+        return (String) searchContextCache.getIfPresent(key);
+    }
+
+    public static void remove(String key) {
```
Comment (mehtaanshul) [inline graphdb/janus/src/main/java/org/apache/atlas/repository/graphdb/janus/SearchContextCache.java:40]: Shouldn't we also remove from searchContextSequenceCache

File: graphdb/janus/src/main/java/org/apache/atlas/repository/graphdb/janus/AtlasElasticsearchQuery.java [hunk 2]
```diff
@@ -194,12 +418,17 @@ private String performDirectIndexQuery(String query, boolean source) throws Atla
         return EntityUtils.toString(response.getEntity());
     }
 
-    private DirectIndexQueryResult getResultFromResponse(String responseString) {
-        DirectIndexQueryResult result = new DirectIndexQueryResult();
-
+    private DirectIndexQueryResult getResultFromResponse(String responseString, boolean async) throws IOException {
```
Comment (mehtaanshul) [inline graphdb/janus/src/main/java/org/apache/atlas/repository/graphdb/janus/AtlasElasticsearchQuery.java:421]: async param is not being used.

File: graphdb/janus/src/main/java/org/apache/atlas/repository/graphdb/janus/AtlasElasticsearchQuery.java [hunk 3]
```diff
@@ -166,6 +170,226 @@ private Map<String, Object> runQueryWithLowLevelClient(String query) throws Atla
...
+        Integer cacheSequenceNumber   = SearchContextCache.getSequence(currentSearchContextId);
+
+        // Check if cache entry exists for the given ID
+        boolean cacheEntryExists = SearchContextCache.get(currentSearchContextId) != null;
+
+        // Handle cases where sequence number is available and greater
+        if (currentSequenceNumber != null && cacheSequenceNumber!= null && currentSequenceNumber > cacheSequenceNumber) {
+            // Sequence number is greater, cancel the request and update cache
+            handleCacheUpdate(currentSearchContextId);
+        } else if (currentSequenceNumber == null || cacheEntryExists) {
```
Comment (mehtaanshul) [inline graphdb/janus/src/main/java/org/apache/atlas/repository/graphdb/janus/AtlasElasticsearchQuery.java:234]: Didn't understand this.

File: graphdb/janus/src/main/java/org/apache/atlas/repository/graphdb/janus/AtlasElasticsearchQuery.java (hunk 2 above)
Comment (sumandas0) [reply graphdb/janus/src/main/java/org/apache/atlas/repository/graphdb/janus/AtlasElasticsearchQuery.java:407]: It is intentional, it is for method overloading in Java

File: graphdb/janus/src/main/java/org/apache/atlas/repository/graphdb/janus/AtlasElasticsearchQuery.java (hunk 3 above)
Comment (sumandas0) [reply graphdb/janus/src/main/java/org/apache/atlas/repository/graphdb/janus/AtlasElasticsearchQuery.java]: We don't need this, removed

File: graphdb/janus/src/main/java/org/apache/atlas/repository/graphdb/janus/SearchContextCache.java (hunk 1 above)
Comment (sumandas0) [reply graphdb/janus/src/main/java/org/apache/atlas/repository/graphdb/janus/SearchContextCache.java:42]: Irrelevant with new cache storing

**Review by mehtaanshul — APPROVED** [review]

**Pull request merged**